### ISSUE TITLE
Yield calculation suggestion

### ISSUE THREAD

**Issue by jessesoslow** [stacking-club/stacking.club#2]
First off, though I'm not sure this is appropriate for a GitHub issue:  Stacking.club is _awesome_.  Very very impressive.

My suggestion is that the yield calculations should be based on the average values over the course of a cycle, and shouldn't remain dynamic based on the value of STX / BTC following the close of the cycle.

I would suggest taking some average value of STX over the period for the cycle and some average value of BTC over the period for the cycle, and calculate yield that wouldn't change following the end of the cycle.  One way to do this would be to take the average of the "close" price for each cycle day for each of STX and BTC.

I think even the yield calculation for a current cycle should use an average.  In particular, I would use a moving average over the course of the current cycle so that the estimated yield for a current cycle is smoothed over the course of a cycle and not as volatile as of the prices of BTC / STX.  In other words, on day two of a cycle, the input for the yield would be the average STX and BTC values for days 1 and 2 of the cycle, day three would use the average values for days 1, 2 and 3, and so on until the last day of the cycle when it takes the average of all cycle days and stops updating the STX / BTC values and the yield remains static from that point forward.  

For the current cycle, one way to do this is by using the current dynamic prices for STX and BTC (say it's day N) averaged with the "close" prices on days 1, 2, ..., N-1.

**Comment by aulneau** [comment]
Hi @jessesoslow, thanks for this, this is really quite helpful for me! I've done a basic implementation of what you've suggested. The calculation is roughly this:


```ts
  const historicalPrices = { stx, btc } // this is the historical price at the day the cycle started

  const cyclesPerYear = 28; // to account for the gaps

  const cyclesMissed = cyclesPerYear / averageCyclesLocked; // averageCyclesLocked = the real average of all stackers locked time

  const cyclesWithRewards = cyclesPerYear - cyclesMissed;

  const costToStackUsd = historicalPrices.stx * minimumThresholdStx; // the price of STX on the start of cycle times the cycle min

  const rewardsUsd = historicalPrices.btc * averageRewardBtc; // the price of BTC at the END of cycle (if cycle is over, else START price)

  const yearlyRewards = rewardsUsd * cyclesWithRewards;

  const apy = yearlyRewards / costToStackUsd; // final apy
```

The main difference with what I have and what you've suggested is that it's only using one point of pricing data (start of cycle pricing).

The APY will no longer fluctuate based on current market data for previous cycles, however, I'd like to add the ability to: 
- get an average of prices throughout the cycle like you mentioned
- we know the exact time folks get rewards, so we can log the price at that time and use that to calculate the APY

**Comment by jessesoslow** [comment]
That's awesome!

Out of curiosity, how did you guys arrive at 28 cycles per year?  My quick math estimates something more like 25.

Here's my calc, assuming a 10 min BTC block time, which might be on the slightly fast side, which might indicate even my estimate is slightly high:

(365 days / year) * (24 hours / day) * (60 min / hour) / (10 min / block) = 52,560 blocks per year

Each cycle is 2100 blocks so number of cycles per year would be:

(52,560 blocks / year) / (2100 blocks / cycle) = 25.0 cycles / year

Am I thinking about this wrong?

**Comment by aulneau** [comment]
Yeah I was wondering about that number too, @hstove came up with it originally, so maybe when he's back from vacation we can understand better how he came upon it.

Your math looks right, and I would actually say that 24 would be pretty accurate in the sense that given the gap between when folks can stack, at most you'd be able to do 12 cycles + 1 gap + 12 cycles = 25 but 24 of them with rewards. Does that seem right to you @jessesoslow?

**Comment by jessesoslow** [comment]
I think you're certainly right in a sense, though I could really go either way on question of which number you should use for the calculation. My slight inclination would be to stick with 25.  My guess is that given the strong community desire to see a change that would allow for continuous stacking, if someone were to lock for 12 cycles today, that change would be implemented before they'd need to take a break for a cycle.  In other words, I'd wager such a person would ultimately be able to continuously stack for 25 cycles.

Also it would save you from having to update when that feature is implemented (which I think is just a question of when, not if).

Though, again, I think either approach would be reasonable.

**Comment by aulneau** [comment]
Yep, that makes sense to me!

I've updated the calculations and seems like we're getting a more correct feeling number. I know this is similar to what OkCoin posted on twitter!

![Screen Shot 2021-02-25 at 8 50 26 AM](https://user-images.githubusercontent.com/11803153/109170676-9e2b8380-7746-11eb-8687-57a196a3549e.png)

**Comment by jessesoslow** [comment]
Awesome!  Yeah, that seems right.

Did something happen to the 1st cycle, though?  Or is the below consistent with OKC's numbers?

![image](https://user-images.githubusercontent.com/37340772/109173354-be147480-7751-11eb-8f49-f0f690fcfcb0.png)

**Comment by aulneau** [comment]
I don't know what OkCoin got for the first cycle, but this thread has some context: https://twitter.com/heynky/status/1363177101965234178

The first cycle had a lot more instances of <4 cycles committed to, which reduces the APY. I don't quite understand exactly all that goes into it, but that seems to be why.

**Comment by jessesoslow** [comment]
Ah, I see.

Seems strange to me that the APY would take into account each individual stacker's choice as to number of cycles stacked.  I would think the APY would be expressing the concept of something like "if you were to put $X into the system and sought to maximize your stacking returns, this is the average APY you would see based on the BTC distributed to Stackers during this cycle."

I guess I see philosophical reason for taking Hank's approach, but I think it kind of conflates some things.  I think a better approach might be to display the APY assuming max cycle participation (25), but allow people to understand what that APY would translate to if they were to only participate in, e.g., every other cycle, or every four cycles, etc.

Also, once continuous stacking is introduced, I'd think this discrepancy would go away.  Even if someone stacked for only 4 cycles, there'd be no reason to not assume they wouldn't stack for the immediate following 4 cycles, and so on.

**Comment by aulneau** [comment]
> I guess I see philosophical reason for taking Hank's approach, but I think it kind of conflates some things. I think a better approach might be to display the APY assuming max cycle participation (25), but allow people to understand what that APY would translate to if they were to only participate in, e.g., every other cycle, or every four cycles, etc.

Yeah this makes a lot of sense. 

I'm curious what you think of this: The current APY math takes into consideration the reward slot value. If we were to give the APY based on full participation, how would you account for if the minimum amount of STX required per slot were to go up over time?

It feels like to me part of why Hank chose to use the real data for cycles stacked is to account for part of this.

Also, all of this is really outside of my expertise lol so thank you so much for this.

**Comment by jessesoslow** [comment]
I think taking into consideration the minimum slot increasing over time is over complicating it.

Personally I would think of the APY as simply the implied earnings over a year _if the conditions of this cycle were to remain the same_.  I'm not sure the APY calculation should be much more complicated than:

[(Value Earned) / (Value Locked)] * 25

**Comment by muneeb-ali** [comment]
Keeping the APY calculation simple makes sense.

It'd also be good to record (a) the average APY per cycle and (b) average USD value of a single reward slot per cycle, and plot these over time i.e., how did the average APY per cycle change and how did the value of a slot (in USD) change over time. For example it's possible that the APY comes down but the USD value per slot goes up.
